File: dart-sdks/lib/src/bcs/reader.dart
import 'dart:typed_data';

import 'uleb.dart';

class BcsReader {
  BcsReader(List<int> bytes) : _bytes = Uint8List.fromList(bytes);

  final Uint8List _bytes;
  int _pos = 0;

  int get position => _pos;

  int remaining() => _bytes.length - _pos;

  int readU8() {
    _ensureInRange(1);
    return _bytes[_pos++];
  }

  int readU16() {
    final bytes = readBytes(2);
    return bytes.buffer.asByteData(bytes.offsetInBytes, 2).getUint16(0, Endian.little);
  }

  int readU32() {
    final bytes = readBytes(4);
    return bytes.buffer.asByteData(bytes.offsetInBytes, 4).getUint32(0, Endian.little);
  }

  int readU64() {
    final bytes = readBytes(8);
    return bytes.buffer.asByteData(bytes.offsetInBytes, 8).getUint64(0, Endian.little);
  }

  Uint8List readBytes(int n) {
    if (n < 0) {
      throw RangeError.value(n, 'n', 'must be non-negative');
    }
    _ensureInRange(n);
    final out = _bytes.sublist(_pos, _pos + n);
    _pos += n;
    return Uint8List.fromList(out);
  }

  bool readBool() {
    final v = readU8();
    if (v == 0) {
      return false;
    }
    if (v == 1) {
      return true;
    }
    throw StateError('invalid bool byte: $v');
  }

  int readUleb128() {
    final decoded = decodeUleb128(_bytes.sublist(_pos));
    _pos += decoded.consumed;
    return decoded.value;
  }

  void _ensureInRange(int len) {
    if (remaining() < len) {
      throw StateError('bcs: out of range');
    }
  }
}

File: py-sdks/src/pysdks/bcs/reader.py
"""BCS byte reader."""

from __future__ import annotations

from dataclasses import dataclass

from .uleb import decode_uleb128


@dataclass
class BCSReader:
    data: bytes
    pos: int = 0

    def remaining(self) -> int:
        return len(self.data) - self.pos

    def read_u8(self) -> int:
        if self.remaining() < 1:
            raise ValueError("bcs: out of range")
        v = self.data[self.pos]
        self.pos += 1
        return v

    def read_u16(self) -> int:
        return int.from_bytes(self.read_bytes(2), "little")

    def read_u32(self) -> int:
        return int.from_bytes(self.read_bytes(4), "little")

    def read_u64(self) -> int:
        return int.from_bytes(self.read_bytes(8), "little")

    def read_bytes(self, n: int) -> bytes:
        if n < 0 or self.remaining() < n:
            raise ValueError("bcs: out of range")
        out = self.data[self.pos : self.pos + n]
        self.pos += n
        return out

    def read_bool(self) -> bool:
        v = self.read_u8()
        if v == 0:
            return False
        if v == 1:
            return True
        raise ValueError(f"invalid bool byte: {v}")

    def read_uleb128(self) -> int:
        value, consumed = decode_uleb128(self.data[self.pos :])
        self.pos += consumed
        return value
